frontend/src/main/scala/akka/viz/frontend/ApiConnection.scala

```diff
@@ -1,44 +1,47 @@
 package akka.viz.frontend
 
-import org.scalajs.dom._
-import org.scalajs.dom.raw.WebSocket
+import org.scalajs.dom
+import org.scalajs.dom.raw._
 
-object ApiConnection {
+import scala.concurrent.Promise
+import scala.concurrent._
+import scala.scalajs.js.JavaScriptException
+import scala.util.{Try, Success}
 
-  trait Upstream {
-    def send(msg: String): Unit
-  }
+trait Upstream {
+  def send(msg: String): Unit
+}
+
+object ApiConnection {
 
-  def apply(url: String, fn: MessageEvent => Unit): Upstream = {
 
-    var webSocket: Option[WebSocket] = None
+  def apply(url: String, fn: MessageEvent => Unit, maxRetries: Int = 1)(implicit ec: ExecutionContext): Future[WebSocket] = {
 
-    //FIXME: implement reconnect and wait for `onopen` event before sending
     def createWebsocket: WebSocket = {
       val ws = new WebSocket(url)
-      ws.onmessage = (messageEvent: MessageEvent) => {
-        fn(messageEvent)
-      }
-      ws.onclose = (c: CloseEvent) => {
-        console.log(c)
-        webSocket = None
-      }
-      ws.onerror = (e: ErrorEvent) => {
-        console.log(e)
-        webSocket = None
-      }
-      webSocket = Some(ws)
       ws
     }
-    webSocket = Some(createWebsocket)
 
-    def getWs = webSocket.getOrElse(createWebsocket)
+    val wsPromise = Promise[WebSocket]()
```

Comment:
WebSocket is callback-based, the only other option would be a busy-wait on WS inside a Future (bad idea in single-threaded code)